File: CPP/BST/101. Symmetric Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isSymmetric(TreeNode* root) {
        if (root->left && root->right){
            return helper(root->left, root->right);
        }
        else if (root->left == nullptr && root->right == nullptr){
            return true;
        }
        return false;
    }

    bool helper(TreeNode* r1, TreeNode* r2){
        if (r1) cout << "r1:" << r1->val;
        if (r2) cout << "r2:" << r2->val << endl;
        if (r1 == nullptr && r2 == nullptr) {
            return true;
        }
        else if (r1 && r2){
            if (r1->val == r2->val) {
                if (helper(r1->left, r2->right) && helper(r1->right, r2->left)){
                    return true;
                }
            }
        }
        return false;
    }
};

File: Python/Binary Tree/101. Symmetric Tree.py
# Definition for a binary tree node.
from typing import Optional
import DS as DS
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
        
class Solution:
    def isSymmetric(self, root: Optional[TreeNode]) -> bool:
        return self.isMirror(root.right, root.left)
        # return root.left root right
        
    def isMirror(self, right, left):
        if right == None and left == None:
            return True
        
        if right and left:
            if right.val != left.val:
                return False
        elif right == None or left == None:
            return False
            
        
        #iter function
        return self.isMirror(right.right, left.left) and self.isMirror(right.left, left.right)
    
        
        
if __name__ == '__main__':
    root = DS.arr2TreeNode([1,2,2,3,4,4,3])
    pos = 1
    sol = Solution()
    print(sol.isSymmetric(root))